bench/simple_consumer_bench.py

```diff
@@ -0,0 +1,111 @@
+import json
+import random
+import subprocess
+import timeit
+
+from pykafka import KafkaClient
+from pykafka.test.utils import get_cluster
+
+
+NUM_PARTITIONS_RANGE = [4, 8, 16, 32, 64]
+MSG_SIZE_BYTES_RANGE = [100, 1000, 10**4, 10**5]
+
+
+def get_topic_name(num_partitions, msg_size_bytes):
+    return "parts{}-size{}".format(num_partitions, msg_size_bytes)
+
+
+def mk_topics(num_partitions_range=NUM_PARTITIONS_RANGE,
+              msg_size_bytes_range=MSG_SIZE_BYTES_RANGE,
+              msg_sum_bytes=10**9,  # limit at reasonable disk footprint
+              max_num_msgs=10**6):  # we won't bench with more than 1e6 msgs
+    cluster = get_cluster()  # make this in advance, like in travis.yml
+    topics = [(np, sz, get_topic_name(np, sz))
+              for np in num_partitions_range
+              for sz in msg_size_bytes_range]
+
+    for num_partitions, msg_size_bytes, topic_name in topics:
+        cluster.create_topic(topic_name, num_partitions, replication_factor=1)
+        print "Producing into {}".format(topic_name)
+        client = KafkaClient(cluster.brokers)
+        n_msgs = min(msg_sum_bytes // msg_size_bytes, max_num_msgs)
+        prod = client.topics[topic_name].get_producer()
+        prod.produce(msg_size_bytes * b" " for _ in xrange(n_msgs))
+
+
+SETUP = ("from pykafka import KafkaClient\n"
+         "from pykafka.rdkafka import RdKafkaSimpleConsumer\n"
+         "from pykafka.common import OffsetType\n"
+         "client = KafkaClient('localhost:9092')\n"
+         "topic = client.topics['{topic_name}']\n"
+         "kwargs = dict(\n"
+         "    auto_offset_reset=OffsetType.EARLIEST,\n"
+         "    consumer_timeout_ms=5000,\n"  # with default (-1) it might hang
+         "    num_consumer_fetchers={num_consumer_fetchers},\n"
+         "    queued_max_messages={queued_max_messages},\n"
+         "    )\n")
+SETUP_FIN = {
+    "pure-py": "cons = topic.get_simple_consumer(**kwargs)",
+    "rdkafka": "cons = RdKafkaSimpleConsumer(topic, client.cluster, **kwargs)",
+    }
+
+
+# We seem to have an issue where consumers don't get garbage-collected, and so
+# their worker threads keep going even after timeit is done with them (and
+# even if we do slip in a stop() call, they don't release the memory for their
+# message queues - I tried that).  As a workaround, we must exit the
+# interpreter after every call to run_bench, and then aggregate the running
+# times in the analysis script instead.
```

Comment:
Just a quick note on this, so I don't forget - I looked at `SimpleConsumer` a bit, and guessed that the gc issue could be due to the worker threads holding a reference to `self` (passed in as a closure). So I had a shot at handing both the fetcher threads and the offset_commit thread a `weakref.proxy` instead, but no luck (even when I removed the `__del__` method (the worker threads, if handed a `weakref.proxy` die anyway, with a `ReferenceError`, so no need for `__del__`).

That is, maybe the above is part of it, but there's also something more that prevents gc.